feat 5340 addition of html and pdf report rendering of encapsulated SR reports.

luissantosHCIT opened the pull request:
<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- 🕮 Read our guide about our Contributing Guide here https://docs.ohif.org/development/contributing -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context

<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->
Fixes #5340 
Requires #5293, #5327, and #5337 before merging.
Commits start August 15th.

This PR brings a React component that can handle SR encapsulated contents.
Currently, the viewer assumes all data in the SR tree are pieces of formatted text.
However, in my environment, I come across SR series that are encapsulated reports.
The encapsulated report is an HTML report in my case.
As part of the component bundle, I included a component for handling PDF if the encountered data is PDF.
The new components default to text like before.

Also, because of the danger of rendering HTML, I added a dependency on `sanitize-html` and take steps to clean the input HTML as much as possible. I make sure no scripts are allowed to execute when viewing HTML reports.

A follow up PR or two will be needed to introduce:

- Refactor the UTF-8 encoding logic to use a more updated version of `dcmjs` library once my PR (455) gets merged.
- Introduce test data and unit tests.

### Changes & Results

<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?
- Before vs After
- Screenshots / GIFs / Videos
-->

- Added HTML rendering component.
- Added PDF rendering component
- Preserved Text Rendering behavior as default.
- Added dependency on `html-sanitize` to help clean up document.
- Added utility function to make sure DICOM inputs are encoded in utf-8.
- Added regexes to make sure we isolate HTML in the payload and discard anything else.
- Corrected logic bug with obtaining the `CodeMeaning` elements in SR nodes.

#### Before

<img width="842" height="323" alt="image" src="https://github.com/user-attachments/assets/09d92e2b-3e11-40b9-94ed-6b9b72aef3d7" />

#### After

<img width="597" height="428" alt="image" src="https://github.com/user-attachments/assets/13e4b667-d9d5-46d1-a0ce-8f5cfffa09e3" />

### Testing

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->
E2E tests and test dataset will need to be introduced in a later PR because I need to figure out how to propose test data.
Also, the PR is getting much bigger than I feel comfortable and I think it will create more burden to the reviewer, so I am breaking this into a few chunks.

TODO: Finalize before and after testing with current test suites in the morning.

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://docs.ohif.org/ -->

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: <!--[e.g. Windows 10, macOS 10.15.4]-->Ubuntu Linux
- [x] Node version: <!--[e.g. 18.16.1]-->20.18.0
- [x] Browser: Firefox 141, Chromium 138.0.7204.157
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

<!-- greptile_comment -->

<h3>Greptile Summary</h3>

This PR adds HTML and PDF rendering support for encapsulated SR reports, replacing the flat text-only `OHIFCornerstoneSRContentItem` with a new component stack (`OHIFCornerstoneSRContainerItem` → `OHIFCornerstoneSREncapsulatedReport` → PDF/HTML/text sub-renderers). It also centralises Blob/MIME handling into `platform/core/src/utils/toBlob.ts` and adds `sanitize-html` for XSS mitigation. The bulk of the previously identified bugs (missing `useEffect` deps, `Buffer` browser crash, `JPEG` undefined variable, null guards, memory leaks, wrong PDF detection, stale `data` closure, and the `measurements` null guard ordering) have all been addressed in the current iteration.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all previously identified P0/P1 bugs have been resolved and only minor style/cleanup findings remain.

The prior round of review surfaced a substantial number of correctness issues. All have been addressed in this iteration. The three remaining comments are P2 style suggestions that do not affect correctness on the happy path.

No files require special attention for merge; OHIFCornerstoneSREncapsulatedReport.tsx and payload.ts are worth a follow-up pass when the test dataset PR lands.

<details open><summary><h3>Vulnerabilities</h3></summary>

- HTML sanitization via `sanitize-html` is in place; `allowVulnerableTags` has been removed and `script` is not in the allowlist.
- `data:` URIs are restricted to `img[src]` only via `allowedSchemesByTag`, mitigating the most common data-URI XSS vector.
- `dangerouslySetInnerHTML` is guarded by `sanitizeHtml()` which calls `extractHTMLFromPayload` + `sanitize-html` before injection.
- The allowlist for tags and attributes is very broad; future tightening is advisable but not a blocking concern.
- No secrets or credentials exposed in the diff.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.tsx | New component handling HTML/PDF/text SR content; previous bugs fixed (useEffect deps, null guard, loading spinner), but stale `mime` closure in the effect is a minor edge-case issue |
| extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedPDFReport.tsx | New PDF viewer component; object URL lifecycle is correctly managed; height=500rem is an invalid HTML attribute value that may cause zero/default height in some browsers |
| extensions/cornerstone-dicom-sr/src/utils/payload.ts | Core payload utilities (MIME detection, HTML sanitization, UTF-8 decode); sanitizer config is verbose with many duplicate entries, but security posture is reasonable; data: URIs restricted to img tag only |
| extensions/cornerstone-dicom-sr/src/utils/srInspection.ts | New utility for inspecting SR tree nodes; null-safety added via optional chaining on ConceptNameCodeSequence |
| extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSRContainerItem.tsx | New container item renderer replacing OHIFCornerstoneSRContentItem; encoding fallback chain (content item to root to default) is correct |
| extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSRMeasurementViewport.tsx | Null guard moved before measurements[measurementSelected] access; encapsulated SR fallback to text viewport added |
| platform/core/src/utils/toBlob.ts | New consolidated blob/MIME utility; b64toBlob backward-compat alias preserved; MimeOptions enum centralises MIME string constants |
| extensions/default/src/DicomWebDataSource/index.ts | Refactored thumbnail src generation to use utils.toBlob and MimeOptions; Jpeg/AcceptJpeg destructured correctly |
| extensions/default/src/utils/getDirectURL.ts | Switches from new Blob(...) to utils.toBlob and from b64toBlob to b64ToBlob; functionally equivalent refactor |
| extensions/cornerstone-dicom-sr/src/utils/constants.ts | Encoding map adds ISO_IR 192 to utf-8; covers the key UTF-8 DICOM encoding; defaults are consistent |

</details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.tsx
Line: 32

Comment:
**Stale `mime` closure in `useEffect` dependency array**

`mime` is used as the fallback argument to `getPayloadType(decoded, mime)` inside the effect, but it is not listed in the dependency array `[props.content, props.expectB64, props.encoding]`. If `props.content` changes between renders and the previous content caused `mime` to be set to a non-default value (e.g. `text/html`), the new content will be evaluated with that stale fallback. In the rare case where the new content cannot be auto-detected, `getPayloadType` would return the stale `mime` instead of the intended default. Adding `mime` to the dependency array (or using the initial `utils.MimeOptions.Default` constant directly as the second argument) prevents this drift.

```suggestion
        const correctMime = getPayloadType(decoded, utils.MimeOptions.Default);
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedPDFReport.tsx
Line: 19

Comment:
**Invalid CSS unit in HTML `height` attribute**

The `height` attribute on `<object>` (and most HTML elements) expects a plain integer (pixels) or a percentage string — not a CSS length with a unit like `rem`. `"500rem"` will be treated as an invalid value by browsers; the effective height will fall back to the browser default, likely making the PDF viewer very short or zero-height. Use a pixel value or control the height via CSS instead.

```suggestion
        <object data={fileURL} width="100%" height="500" type="application/pdf">
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: extensions/cornerstone-dicom-sr/src/utils/payload.ts
Line: 35-45

Comment:
**Duplicate entries in `allowedTags` and `nonBooleanAttributes` / `allowedAttributes`**

Many tags (e.g. `h1`–`h4`, `li`, `ul`, `ol`, `table`, `tr`, `td`, `tbody`, `th`, `a`, `img`, `br`, `em`, `strong`, `i`, `sub`, `sup`, `p`) appear two or three times in `allowedTags`. Likewise `style`, `href`, `height`, `width`, `class`, `id`, `src`, `name`, `content`, `colspan`, `rowspan`, `http-equiv` all appear twice in both `nonBooleanAttributes` and the `allowedAttributes['*']` array. While `sanitize-html` deduplicates internally, the redundancy makes this configuration harder to audit for security correctness. Removing duplicates would reduce the risk of unintended entries being missed during future review.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (12): Last reviewed commit: ["fix(dicom-sr) fixed a potential React re..."](https://github.com/ohif/viewers/commit/7395ffd5beb9668c1f63b6d7cfa4d98a309e441f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=23872630)</sub>

<!-- /greptile_comment -->

Commit:
fix Added new React components to handle different types of encapsulated report contents.
Basically, the viewer can now display HTML, Markdown, PDF, and Text reports with the appropriate page object.
Added sanitize-html, Markdown, and @types/sanitize-html as dependencies.
TODO: Add proper HTML sanitization options so we can preserve the core portions of the document but remove anything that can cause security issues.
TODO: Hook the html sanitization routine into the component logic.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore simplified report component logic.
Adjusted the html extraction regex to ensure we capture html contents only.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore refactored file naming and removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore refactored file naming again now that I understand the higher levels of abstraction better.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
fix The viewport was not displaying the Code meaning properly because it was not meeting the standard specification of looking into the ConceptNameCodeSequence array. It assumed that the CodeMeaning field was at the array object level.
Refactored code for clarity and added type definitions.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
fix Improper selection of the code meaning field from the conceptnamecodesequence node. This is the same bug as at the container level.
Minor API refactor for clarity.
Restored logic from initial commit since it handles the continuity of content which I typically don't and I want ot avoid breaking someone else's expectations. If current solution turns out to not be complete for embedded reports, I shall revisit this file.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
feat added html sanitizer options. Added temporary utf-8 encoding function for enforcing utf-8 encoding in payload. Updated the sanitizeHTML function to go through all steps necessary to end up with a clean HTML document.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
feat added specific character set field to interface type. The goal is to track the specified encoding type so that we can convert to utf-8 accordingly

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
feat hooked the new html sanitization logic to the report component so that we can guarantee that PACS or VNA payloads are safe to render.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
feat added tracking of initial encoding in viewer port so that we can render encapsulated reports correctly.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore added default to state

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

Commit:
chore added document strings to the payload module.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

luissantosHCIT commented:
I am not sure what Cypress is complaining about.
The test suite runs without errors in my local environment and I have not changed the tests at all.

<img width="1559" height="1407" alt="image" src="https://github.com/user-attachments/assets/3674b95d-3bd4-4fa8-a609-7f140df7e9cf" />

wayfarer3130 commented:
Can you provide an anonymized example of both pdf and html reports which can be added to the automated test data set?  I can get those uploaded.
You must include a notification in that which indicates there is no PHI data in them, and that OHIF has permission to include and distribute them in the viewer test dataset.

wayfarer3130 commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,492 @@
+ * @param {string} mime MIME to add to Blob so other components can know how to handle contents.
+ * @return Blob
+ */
+export function stringToBlob(data: string, mime: string = payloadMIMEOptions.DEFAULT): Blob {
```
This should be a shared external function.  It gets repeated several times in the code so sharing it means any bugs in it could be fixed.

luissantosHCIT replied:
Sorry, you are probably asking about a TypeScript specific thing. I am not understanding this request. The function is already exported from the payload module in utils. I thought I had all areas pointing to this function as an import. Is it that I left places accidentally duplicating code?

My brain cleans the code in passes and sometimes I forget to do one last pass.

luissantosHCIT replied:
Hmm, I still do not understand this one. Are you asking me to move this function somewhere else? Where were you thinking this function should live?

wayfarer3130 commented on extensions/cornerstone-dicom-sr/src/utils/srInspection.ts:
```diff
@@ -0,0 +1,35 @@
+): string {
+  let item: ConceptNameCodeSequenceItem = conceptNameCodeSequence[0];
+  const { CodeMeaning } = item;
+  return CodeMeaning ?? "";
```
Why not 
{
   return conceptNameCodeSequence?.[0]?.CodeMeaning || ""
}

Really, all you are doing is extracting path values, and a single extractor is easier to manage.

luissantosHCIT commented on extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.tsx:
```diff
@@ -0,0 +1,64 @@
+import React, { useEffect } from 'react';
+import Markdown from 'marked-react';
```
Thank you. I will read more about this and educate myself and then submit a commit.

luissantosHCIT committed:
Merge remote-tracking branch 'origin/5340_addition_of_html_report_rendering' into 5340_addition_of_html_report_rendering

# Conflicts:
#	extensions/cornerstone-dicom-sr/package.json

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
fix version string of dicom sr dependencies

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
chore cleanup of code, addition of null checks, and tightening of potential XSS. Keep in mind that we do need to be able to load inlined image blobs from images for encapsulated HTML reports out there.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore renaming variable.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore adding const.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore simplified return extraction.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix making marked-react lazy load.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
5340 addition of html report rendering (#3)

* fix Added new React components to handle different types of encapsulated report contents.
Basically, the viewer can now display HTML, Markdown, PDF, and Text reports with the appropriate page object.
Added sanitize-html, Markdown, and @types/sanitize-html as dependencies.
TODO: Add proper HTML sanitization options so we can preserve the core portions of the document but remove anything that can cause security issues.
TODO: Hook the html sanitization routine into the component logic.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore simplified report component logic.
Adjusted the html extraction regex to ensure we capture html contents only.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore refactored file naming and removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore refactored file naming again now that I understand the higher levels of abstraction better.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix The viewport was not displaying the Code meaning properly because it was not meeting the standard specification of looking into the ConceptNameCodeSequence array. It assumed that the CodeMeaning field was at the array object level.
Refactored code for clarity and added type definitions.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix Improper selection of the code meaning field from the conceptnamecodesequence node. This is the same bug as at the container level.
Minor API refactor for clarity.
Restored logic from initial commit since it handles the continuity of content which I typically don't and I want ot avoid breaking someone else's expectations. If current solution turns out to not be complete for embedded reports, I shall revisit this file.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added html sanitizer options. Added temporary utf-8 encoding function for enforcing utf-8 encoding in payload. Updated the sanitizeHTML function to go through all steps necessary to end up with a clean HTML document.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added specific character set field to interface type. The goal is to track the specified encoding type so that we can convert to utf-8 accordingly

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat hooked the new html sanitization logic to the report component so that we can guarantee that PACS or VNA payloads are safe to render.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added tracking of initial encoding in viewer port so that we can render encapsulated reports correctly.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore added default to state

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore added document strings to the payload module.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix version string of dicom sr dependencies

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore cleanup of code, addition of null checks, and tightening of potential XSS. Keep in mind that we do need to be able to load inlined image blobs from images for encapsulated HTML reports out there.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore renaming variable.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore adding const.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore simplified return extraction.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* fix making marked-react lazy load.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

---------

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>
Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>
Co-authored-by: Luis M. Santos <luis.santos2@nih.gov>

luissantosHCIT committed:
fix(SR) missing measurement check found in the MeasurementSR component because of another bug with SR component selection.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) added default return of OHIFCornerstoneSRTextViewport in the event that a report self identifies as an SR Measurement Report but actually having an encapsulated report instead. It's a real thing.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) added CSS to invert element color so reports (which are typically white or without the color set with black text have its text pop and match the viewer's theme. There could be occasional defects since this is based on reports expected in my use case.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'patched' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
5340 addition of html report rendering (#4)

* fix Added new React components to handle different types of encapsulated report contents.
Basically, the viewer can now display HTML, Markdown, PDF, and Text reports with the appropriate page object.
Added sanitize-html, Markdown, and @types/sanitize-html as dependencies.
TODO: Add proper HTML sanitization options so we can preserve the core portions of the document but remove anything that can cause security issues.
TODO: Hook the html sanitization routine into the component logic.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore simplified report component logic.
Adjusted the html extraction regex to ensure we capture html contents only.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore refactored file naming and removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore refactored file naming again now that I understand the higher levels of abstraction better.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix The viewport was not displaying the Code meaning properly because it was not meeting the standard specification of looking into the ConceptNameCodeSequence array. It assumed that the CodeMeaning field was at the array object level.
Refactored code for clarity and added type definitions.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix Improper selection of the code meaning field from the conceptnamecodesequence node. This is the same bug as at the container level.
Minor API refactor for clarity.
Restored logic from initial commit since it handles the continuity of content which I typically don't and I want ot avoid breaking someone else's expectations. If current solution turns out to not be complete for embedded reports, I shall revisit this file.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore removed no longer needed code.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added html sanitizer options. Added temporary utf-8 encoding function for enforcing utf-8 encoding in payload. Updated the sanitizeHTML function to go through all steps necessary to end up with a clean HTML document.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added specific character set field to interface type. The goal is to track the specified encoding type so that we can convert to utf-8 accordingly

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat hooked the new html sanitization logic to the report component so that we can guarantee that PACS or VNA payloads are safe to render.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* feat added tracking of initial encoding in viewer port so that we can render encapsulated reports correctly.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore added default to state

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* chore added document strings to the payload module.

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>

* fix version string of dicom sr dependencies

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore cleanup of code, addition of null checks, and tightening of potential XSS. Keep in mind that we do need to be able to load inlined image blobs from images for encapsulated HTML reports out there.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore renaming variable.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore adding const.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* chore simplified return extraction.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* fix making marked-react lazy load.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* fix(SR) missing measurement check found in the MeasurementSR component because of another bug with SR component selection.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* fix(SR) added default return of OHIFCornerstoneSRTextViewport in the event that a report self identifies as an SR Measurement Report but actually having an encapsulated report instead. It's a real thing.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

* fix(SR) added CSS to invert element color so reports (which are typically white or without the color set with black text have its text pop and match the viewer's theme. There could be occasional defects since this is based on reports expected in my use case.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

---------

Signed-off-by: Luis M. Santos <luis.santos2@nih.gov>
Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>
Co-authored-by: Luis M. Santos <luis.santos2@nih.gov>

wayfarer3130 commented on extensions/cornerstone-dicom-sr/src/components/OHIFLazyMarkdown.tsx:
```diff
@@ -0,0 +1,20 @@
+import React, { lazy, Suspense } from 'react';
+
+// Since I was not sure how to approach lazy loading of components, I consulted with Google Search/ Gemini.
```
You can remove this comment - Just the implementation is ok for now, and that is self explanatory.

wayfarer3130 commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,495 @@
+ */
+export const htmlSanitizerOptions = {
+  allowVulnerableTags: true, //We block most styles below, but we should allow some styles through for now so documents can display.
+  allowedTags: [
```
Consider a static string for allowedTags with it comma separated, and then a getter method for allowedTags

get allowedTags() {
   return this._allowedTags || (this._allowedTags = allowedTagsComma.split(','));
}

That will be a bit slower first time access, but is much more efficient for startup time.  The number of elements is just about big enough to be worthwhile.

luissantosHCIT replied:
Would you be kind to elaborate? This is a struct following the sanitizeHTML package expected object syntax. How would the struct have the array initialized by the time it gets passed to the sanitization function?

Thank you!

wayfarer3130 commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,495 @@
+ * @return string
+ */
+export function getPayloadType(payload: string, suggestedMime: string = 'text/plain') {
+  // PDF
```
Since you are working with a custom function, it is worth splitting the payload into the first 2k characters, or the whole string if less than, say 32k.
const prefix = payload.length < 25k ? payload : payload.substring(0,2048);

and then test on prefix - otherwise for a large payload, this gets really slow.

luissantosHCIT committed:
chore(SR) removed unnecessary comment.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) Added comment reminding us to move srInspection into @cornerstonejs/metadata.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) Allowed extractHTMLFromPayload to throw exception if the payload lacks any valid html contents.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

wayfarer3130 committed:
Updated lockfile

luissantosHCIT committed:
refactor(SR) defined toBlob family of functions in @ohif/core. Collected commonly used mimes into MimeOptions object (it was an enum but converted to object for easier export).

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
refactor(SR) changed usages of new Blob to simply use the new toBlob or csvToBlob or base64ToBlob functions as necessary.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
refactor(SR) removal of stringToBlob in favor of the new toBlob function.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,495 @@
+  if (results && results.length) {
+    return results.shift();
+  }
+  return data;
```
Do you know why a JavaScript regex might sporadically fail?

I think I had the guard because of such an issue and I probably forgot (I was working on multiple projects back then).

<img width="1040" height="986" alt="Image" src="https://github.com/user-attachments/assets/3103c6eb-2ad3-4f67-9961-bc4a971481e6" />

I know React runs renders multiple times (at least in dev mode). However, I do not know how that interacts with the RegEx engine since the input is the same and ran against it 3 times in my example above and the 2nd attempt is the one yielding a `null` object. 

For now, what I am going to do is attempt the shift(), catch the exception, and then default to an empty string. Let me see if that allows us to block legitimately bad inputs while still being able to display the report (in dev mode at least).

luissantosHCIT committed:
fix(SR) removal of g flag in regex causing failure due to React's multiple render passes in the dev context.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) cleaned up the object and added a Spinner if we are still waiting on processing the payload.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) limited the sample of the payload used for testing the mime of the payload.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) made sure extracting the HTML defaults to an empty string.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) added unit tests to the SR component's utils.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) added test for sanitizeHTML

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
style(SR) renamed sanitizeHTML to sanitizeHtml for style consistency.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) updated the HTML regex to handle a large number of scenarios exposed while building the unit tests.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) adjusted unit tests after changes to HTML regex.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge remote-tracking branch 'origin/5340_addition_of_html_report_rendering' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
fix(dependency) updated dependency as reported by audit.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(lockfile) updated lockfiles after updating dependencies

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(lockfile) updated bun lockfile

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(SR) Renamed constant to NON_RENDERABLE_MODALITIES to better reflect which modalities we cannot show via the basic viewer and removed SR from that list of unrenderable modalities.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) Cleaned up unused CSS

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
chore(SR) converted CSS to TailwindCSS and inlined.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'patched' into 5340_addition_of_html_report_rendering

# Conflicts:
#	extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.module.css
#	extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.tsx

luissantosHCIT committed:
chore(SR) forgot to remove no longer necessary css module.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
Merge branch 'OHIF:master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,468 @@
+ * Set of options to pass through to sanitize-html.
+ *
+ * Over time, we should optimize this structure towards maximum security possible. The idea is to
+ * allow some level of document fidelity as generated by vendors but help guard users from any potential
+ * malicious payloads that might be living in the user's data repository (e.g. VNA).
+ */
```
Hmm, that is an interesting suggestion. I am going to look further into this to figure out the full range of styles I can passthrough. 

A little background here. The changes I originally proposed enables the rendering of encapsulated HTML (primarily) reports as Philips had them in the CareStream PACS. I was hoping to eliminate vectors of vulnerability from arbitrary payloads with `sanitize-html`. However, a full sanitization of the html would strip styling and a few urls which would break document fidelity. At the time, the only way I knew without manually having to specify each style and edge case was by allowing vulnerable tags through. 

I agree with greptile here because I didn't mean for script tags to make it through, but I was prioritizing proper rendering of the document.

luissantosHCIT replied:
I ended up removing the `allowVulnerableTags` option but kept the `style` tag as one of the whitelisted tags because some documents do need styles present in the global style tag.

luissantosHCIT committed:
fix(bug) rename JPEG => Jpeg which happened in 6 lines I missed prior to submitting work on a refactor.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(bug) incorporated the UTF-8 conversion directly since Buffer is a NodeJS only API.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(bug) corrected no longer valid null check.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(bug) removed the allowVulnerableTags option to html-sanitize. Style tag was kept because some documents do rely on global styling.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(bug) corrected the order in which I perform a guard check. I did not realize that the variable was being initialized before a guard I added in prior commits.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(tests) missed one line that could use the MimeOptions enum.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(tests) corrected mistake present in test prior to my stylistic refactor.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) added null to expected textContent type.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT commented on extensions/cornerstone-dicom-sr/src/utils/payload.ts:
```diff
@@ -0,0 +1,469 @@
+    'img',
+    'sup',
+    'sub',
+    'style',
```
The viewer uses version `2.17.0` and it does not appear to silently remove the style tag. I confirmed by loading the website with no-cache enabled and opened up a report that contains the `<style>` tag without issues. I inspected the injected HTML and the style tag was present in the `<head>` section.

luissantosHCIT committed:
fix(dicom-sr) added iso ir 192 to utf-8 mapping in the encodings. This will be better handled once encoding improvements are merged into dcmjs.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) corrected placement of useEffect's array. Oops.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge branch 'master' into 5340_addition_of_html_report_rendering

luissantosHCIT committed:
fix(dicom-sr) removed setMime and such from useEffect array.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) make sure to recreate the blob with the identified mime before passing to (potentially) the PDF component.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
Merge remote-tracking branch 'origin/5340_addition_of_html_report_rendering' into 5340_addition_of_html_report_rendering

luissantosHCIT commented on extensions/cornerstone-dicom-sr/src/components/OHIFCornerstoneSREncapsulatedReport.tsx:
```diff
@@ -0,0 +1,73 @@
+          <div dangerouslySetInnerHTML={{ __html: sanitizeHtml(textContent) }} />
+        </blockquote>
+      );
+    case utils.MimeOptions.Pdf:
+      return (
+        <blockquote>
+          <OHIFCornerstoneSREncapsulatedPDFReport content={data} />
+        </blockquote>
+      );
```
It is best to make this change at the time we identify the correct mime during the previous `useEffect`. That's what I did. Otherwise, good suggestion.

luissantosHCIT committed:
fix(dicom-sr) moved constants to its own source file in utils.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) added function for checking the encoding as prescribed by the header.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) added props fields/interface and passthrough logic for accepting the last global encoding if the current node does not switch encodings. It allows for a node root to dictate the encoding of a subtree.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) added backwards compatibility export of b64ToBlob.

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>

luissantosHCIT committed:
fix(dicom-sr) fixed a potential React re-fire?

Signed-off-by: Luis M. Santos <lsantos@medicalmasses.com>
